tty/zynq7000-uart/zynq7000-uart.c

```diff
@@ -0,0 +1,498 @@
+{
+	msg_t msg;
+	uart_t *uart;
+	unsigned long int rid;
+	uint32_t port = (uint32_t)arg;
+
+	for (;;) {
+		if (msgRecv(port, &msg, &rid) < 0)
+			continue;
+
+		switch (msg.type) {
+			case mtOpen:
+				break;
+
+			case mtClose:
+				break;
+
+			case mtWrite:
+				if ((uart = uart_get(&msg.i.io.oid)) == NULL) {
+					msg.o.io.err = -ENOENT;
+					break;
+				}
+
+				msg.o.io.err = libtty_write(&uart->tty, msg.i.data, msg.i.size, msg.i.io.mode);
+				break;
+
+			case mtRead:
+				if ((uart = uart_get(&msg.i.io.oid)) == NULL) {
+					msg.o.io.err = -ENOENT;
+					break;
+				}
+
+				msg.o.io.err = libtty_read(&uart->tty, msg.o.data, msg.o.size, msg.i.io.mode);
+				break;
+
+			case mtGetAttr:
+				if (msg.i.attr.type == atPollStatus && (uart = uart_get(&msg.i.io.oid)) != NULL) {
+					msg.o.attr.val = libtty_poll_status(&uart->tty);
+					msg.o.attr.err = EOK;
+					break;
+				}
+
+				msg.o.attr.err = -EINVAL;
+				break;
+
+			case mtDevCtl:
+				uart_ioctl(port, &msg);
+				break;
+		}
+
+		msgRespond(port, &msg, rid);
+	}
+}
+
+
+static void uart_mkDevFiles(void)
+{
+	int i;
+	oid_t dir;
+	char path[12];
+
+	while (lookup("/", NULL, &dir) < 0)
+		usleep(10000);
+
+	for (i = 0; i < UARTS_MAX_CNT; ++i) {
+		if (uart_common.uarts[i].base == NULL)
+			continue;
+
+		snprintf(path, sizeof(path), "/dev/uart%d", i);
+		if (create_dev(&uart_common.uarts[i].oid, path) < 0)
+			debug("uart: Cannot create device file.\n");
+	}
+}
+
+
+static int uart_setPin(uint32_t pin)
+{
+	platformctl_t ctl;
+
+	ctl.action = pctl_set;
+	ctl.type = pctl_mio;
+
+	/* Set default properties for UART's pins */
+	ctl.mio.pin = pin;
+	ctl.mio.l0 = ctl.mio.l1 = ctl.mio.l2 = 0;
+	ctl.mio.l3 = 0x7;
+	ctl.mio.speed = 0;
+	ctl.mio.ioType = 1;
+	ctl.mio.pullup = 0;
+	ctl.mio.disableRcvr = 0;
+
+	switch (pin) {
+		/* Uart Rx */
+		case pctl_mio_pin_10:
+		case pctl_mio_pin_49:
+			ctl.mio.triEnable = 1;
+			break;
+
+		/* Uart Tx */
+		case pctl_mio_pin_11:
+		case pctl_mio_pin_48:
+			ctl.mio.triEnable = 0;
+			break;
+
+		default:
+			return -EINVAL;
+	}
+
+	return platformctl(&ctl);
+}
+
+
+static int uart_initAmbaClk(unsigned int dev)
+{
+	platformctl_t ctl;
+
+	ctl.action = pctl_set;
+	ctl.type = pctl_ambaclock;
+
+	ctl.ambaclock.dev = dev;
+	ctl.ambaclock.state = 1;
+
+	return platformctl(&ctl);
+}
+
+
+static int uart_initClk(void)
+{
+	platformctl_t ctl;
+
+	ctl.action = pctl_set;
+	ctl.type = pctl_devclock;
+
+	/* Set IO PLL as source clock and set divider:
+	 * IO_PLL / 0x14 :  1000 MHz / 20 = 50 MHz     */
+	ctl.devclock.dev = pctl_ctrl_uart_clk;
+	ctl.devclock.clkact0 = 0x1;
+	ctl.devclock.clkact1 = 0x1;
+	ctl.devclock.srcsel = 0;
+	ctl.devclock.divisor0 = 0x14;
+
+	return platformctl(&ctl);
+}
+
+
+static void uart_destroy(unsigned int n)
+{
+	uart_t *uart = &uart_common.uarts[n];
+
+	munmap((void *)uart->base, _PAGE_SIZE);
+	libtty_close(&uart->tty);
+	libtty_destroy(&uart->tty);
+	resourceDestroy(uart->cond);
+	resourceDestroy(uart->lock);
```

Comment:
You're removing resource number 0, if we haven't created it already. So, you might init uart0 (uart->cond == 0) and then fail to init uart1 - and remove cond of uart0. That's why I don't like this function ;)

Reply:
Yeah, you have convinced me. Removed `uart_destroy`.